### PR TITLE
ref(ratelimits): surface concurrent requests to access logs

### PR DESCRIPTION
**Pull request by volokluev**
### Context
All context on this project and reasoning for the implementation can be found [here](https://www.notion.so/sentry/Concurrent-Rate-Limiter-ebd5b7562aed4d01b102d83774a06b11)

This is the first step of concurrent rate limits in sentry. In order to understand what to rate limit by, we fist need to understand what the concurrent usage of our API is. 

### This PR does the following:

* Implements the concurrent rate limiting mechanism which will be used in sentry
* Surfaces the data to the access logs
* Does NOT enforce any concurrent rate limits

Some unrelated but otherwise important things:
* Makes the rate limit middleware fail open. As it is, an exception in that middleware would shut down the request

### What is needed from reviewers

* Are there any test cases that I missed?
* Is it clear what I'm testing and why?
* Would you know how to enable concurrent rate limiting if you had to?
* Does the implementation of the concurrent rate limiter make sense?

### CODE REVIEW

File: src/sentry/middleware/access_log.py [hunk 1]
```diff
@@ -38,6 +38,23 @@ def _get_token_name(auth: RequestAuth) -> str | None:
...
+        "concurrent_limit": str(None),
+        "concurrent_requests": str(None),
+    }
+
+    rate_limit_metadata = getattr(request, "rate_limit_metadata", None)
+    if not rate_limit_metadata:
+        return default
+    res = {}
+    for field in default:
+        res[field] = str(getattr(rate_limit_metadata, field, None))
```
Comment (AniketDas-Tekky) [inline src/sentry/middleware/access_log.py:54]: `rate_limit_type` will never return the default value of `ERROR`. is this expected?

Comment (volokluev) [reply src/sentry/middleware/access_log.py:54]: If there is no rate limit metadata on the request, it will return `ERROR` otherwise it will be `None`.

File: src/sentry/middleware/ratelimit.py [hunk 2]
```diff
@@ -19,42 +27,55 @@ class RatelimitMiddleware(MiddlewareMixin):
...
+        rate_limit_metadata: RateLimitMeta | None = getattr(request, "rate_limit_metadata", None)
+        if rate_limit_metadata:
+            response["X-Sentry-Rate-Limit-Remaining"] = rate_limit_metadata.remaining
+            response["X-Sentry-Rate-Limit-Limit"] = rate_limit_metadata.limit
+            response["X-Sentry-Rate-Limit-Reset"] = rate_limit_metadata.reset_time
+            response[
+                "X-Sentry-Rate-Limit-ConcurrentRemaining"
+            ] = rate_limit_metadata.concurrent_remaining
+            response["X-Sentry-Rate-Limit-ConcurrentLimit"] = rate_limit_metadata.concurrent_limit
+        finish_request(request.rate_limit_key, request.rate_limit_uid)
```
Comment (AniketDas-Tekky) [inline src/sentry/middleware/ratelimit.py:80]: If for someone reason, `finish_request` doesn't run (like from a crash), would the uid leak?

File: src/sentry/middleware/access_log.py (hunk 1 above)
Comment (volokluev) [reply src/sentry/middleware/access_log.py:54]: I can make `ERROR` more descriptive and call it `NORATELIMITMETADATA` instead maybe?

Comment (AniketDas-Tekky) [reply src/sentry/middleware/access_log.py:54]: That could work, or something like `DNE`

File: src/sentry/ratelimits/utils.py [hunk 3]
```diff
@@ -27,6 +28,15 @@
     "members:org-invite-to-email": {"limit": 10, "window": 3600 * 24},
 }
 
+_CONCURRENT_RATE_LIMITER = ConcurrentRateLimiter()
+
+
+def concurrent_limiter() -> ConcurrentRateLimiter:
```
Comment (AniketDas-Tekky) [inline src/sentry/ratelimits/utils.py:34]: Why is this needed?

File: src/sentry/ratelimits/utils.py [hunk 4]
```diff
@@ -120,21 +130,44 @@ def get_rate_limit_value(
...
+    # TODO: This is not as performant as it could be. The roundtrip betwwen the server and redis
+    # is doubled because the fixd window limit and concurrent limit are two separate things with different
+    # paths. Ideally there is just one lua script that does both and just says what kind of limit was hit
+    # (if any)
+    rate_limit_type = RateLimitType.NOT_LIMITED
+    window_limited, current, reset_time = ratelimiter.is_limited_with_value(
         key, limit=rate_limit.limit, window=rate_limit.window
     )
-    remaining = rate_limit.limit - current if not is_limited else 0
+    remaining = 0
```
Comment (AniketDas-Tekky) [inline src/sentry/ratelimits/utils.py:142]: Don't think you need this line

**Review by AniketDas-Tekky — COMMENTED** [review]
Overall LGTM. Just a few questions.

What would the rate limit type be if a request was both fixed window limited and concurrent limited in the same request?

File: src/sentry/ratelimits/utils.py (hunk 3 above)
Comment (volokluev) [reply src/sentry/ratelimits/utils.py:34]: It's a best practice to put global state behind a function call.

**Comment by volokluev** [comment]
> What would the rate limit type be if a request was both fixed window limited and concurrent limited in the same reques

Fixed window takes priority

File: src/sentry/scripts/ratelimits/api_limiter.lua [hunk 5]
```diff
@@ -0,0 +1,27 @@
+-- TODO: This is currently just the concurrent rate limiter however this should subsume the fixed
```
Comment (volokluev) [inline src/sentry/scripts/ratelimits/api_limiter.lua:1]: TODO: make sure this is well explained

File: src/sentry/scripts/ratelimits/api_limiter.lua [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+--  concurrent_limit, request_uid, current_time, max_tll_seconds
+--
+-- Output:
+-- current_executions (including the one that was just added), request_allowed?
+local key = KEYS[1]
+
+local concurrent_limit = tonumber(ARGV[1])
+local request_uid = ARGV[2]
+local cur_time = tonumber(ARGV[3])
+local max_tll_seconds = tonumber(ARGV[4])
```
Comment (volokluev) [inline src/sentry/scripts/ratelimits/api_limiter.lua:17]: TODO: comment this better for someone who does not know redis already

File: src/sentry/ratelimits/concurrent.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    def namespaced_key(self, key: str) -> str:
+        return f"concurrent_limit:{key}"
+
+    def start_request(self, key: str, limit: int, request_uid: str) -> ConcurrentLimitInfo:
+        redis_key = self.namespaced_key(key)
+        try:
+            res = rate_limit_info(
+                self.client, [redis_key], [limit, request_uid, time(), self.max_ttl_seconds]
+            )
+            current_executions, request_allowed = res
```
Comment (volokluev) [inline src/sentry/ratelimits/concurrent.py:49]: remove this line

File: tests/sentry/ratelimits/test_redis_concurrent.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+from sentry.ratelimits.concurrent import DEFAULT_MAX_TTL_SECONDS, ConcurrentRateLimiter
+from sentry.testutils import TestCase
+
+
+class ConcurrentLimiterTest(TestCase):
+    def setUp(self):
+        self.backend = ConcurrentRateLimiter()
+
+    def test_add_and_remove(self):
```
Comment (ceorourke) [inline tests/sentry/ratelimits/test_redis_concurrent.py:16]: Can you add a docstring to this test?

File: tests/sentry/ratelimits/test_redis_concurrent.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        limit = 8
+        with freeze_time("2000-01-01"):
+            for i in range(1, limit + 1):
+                assert (
+                    self.backend.start_request("foo", limit, f"request_id{i}").current_executions
+                    == i
+                )
+
+            info = self.backend.start_request("foo", limit, "request_id_12")
```
Comment (ceorourke) [inline tests/sentry/ratelimits/test_redis_concurrent.py:25]: Is the `request_id` number 12 just because that's higher than the `limit` and you want it to be over the limit?

File: src/sentry/middleware/ratelimit.py (hunk 2 above)
Comment (volokluev) [reply src/sentry/middleware/ratelimit.py:85]: yes it would, but it would be cleaned up within 30 seconds

File: tests/sentry/ratelimits/test_redis_concurrent.py (hunk 8 above)
Comment (volokluev) [reply tests/sentry/ratelimits/test_redis_concurrent.py:16]: done, let me know if it makes sense now

File: tests/sentry/ratelimits/test_redis_concurrent.py (hunk 9 above)
Comment (volokluev) [reply tests/sentry/ratelimits/test_redis_concurrent.py]: correct. I added a comment. let me know if that's more clear

**Review by AniketDas-Tekky — APPROVED** [review]

File: src/sentry/middleware/access_log.py [hunk 10]
```diff
@@ -38,6 +38,23 @@ def _get_token_name(auth: RequestAuth) -> str | None:
     return token_class.__name__ if token_class else None
 
 
+def _get_rate_limit_stats_dict(request: Request) -> dict[str, str]:
+    # TODO:: plumb the rate limit group up here as well for better future analysis
+    default = {
+        "rate_limit_type": "DNE",
```
Comment (ceorourke) [inline src/sentry/middleware/access_log.py:44]: What does DNE stand for?

Comment (volokluev) [reply src/sentry/middleware/access_log.py:44]: does not exist

**Review by ceorourke — APPROVED** [review]
